[Sherman Ng Wei Sheng] iP

shermannws opened the pull request:
# Duke

> The ~Fake~ **Superior** Duke :sunglasses:
> More emojis can be found [here](https://github.com/ikatyang/emoji-cheat-sheet/blob/master/README.md)

Duke is a CLI task manager application that allows you to unleash your productivity potentials within you. It is,
* Text-based
* User Friendly
* Super fast to use

## Commands that Duke understands

1. **list** *(list out all the outstanding task)*
2. **todo [description]** *(add a todo task)*
3. **deadline [description] /by [yyyy-mm-dd]** *(add a deadline task)*
4. **event [description] /at [yyyy-mm-dd]** *(add an event task)*
5. **done [index]** *(mark task at given index as done)*
6. **delete [index]** *(removes task at given index)*
7. **find [search string]** *(find all corressponding tasks that contains that string)*

## Running Duke

If you Java programmer, you can use it to practice Java too. Here's the main method:
```java
public static void main(String[] args) {
    Duke duke = new Duke("data/duke.txt");
    duke.start();
}
```
You can also use the JAR file directly!
1. Copy the jar file into an empty folder
2. Open a command window in that folder
3. Run the command `java -jar Duke.jar`

## iP Progress
- [x] Basic Functionalities of Duke
- [x] Javadoc
- [x] Unit & Regression Testing
- [x] Gradle
- [x] GUI

Commit:
Add Gradle support

shermannws committed:
Implement Level-1 with Javadoc

shermannws committed:
Implement Level-2 with Javadoc

shermannws committed:
Implement Level-3 with Javadoc

shermannws committed:
Implement Level-4

shermannws committed:
Implement Level-4

shermannws committed:
Implement A-TextUiTesting

shermannws committed:
Implement Level-5, A-Exceptions, Refactor Code, Update Javadoc

shermannws committed:
[Update] Implement Level-5, A-Exceptions, Refactor Code, Update Javadoc

shermannws committed:
Implement Level-6, A-Collections, Refactor Code and Updated A-TextUiTesting

shermannws committed:
Update Javadoc according to conventions

shermannws committed:
Implement Level-7 saving data feature

shermannws committed:
Remove redundant import statement

shermannws committed:
Implement Level-8

shermannws committed:
Merge branch 'branch-Level-7'

shermannws committed:
Merge branch 'branch-Level-8'

shermannws committed:
Update regression testing script

shermannws committed:
Implement A-MoreOOP

shermannws committed:
Implement A-Packages

shermannws committed:
Implement A-JUnit

shermannws committed:
Implement A-Jar

shermannws committed:
Implement A-JavaDoc

shermannws committed:
Implement A-CodingStandard

shermannws committed:
Implement Level-9

shermannws committed:
Merge branch 'branch-A-JavaDoc'

shermannws committed:
Merge branch 'branch-A-CodingStandard'

shermannws committed:
Merge branch 'branch-Level-9'

shermannws committed:
Merge remote-tracking branch 'origin/add-gradle-support' into branch-A-Gradle

ZhaoPeiduo commented on src/main/java/duke/MarkDoneCommand.java:
```diff
@@ -0,0 +1,54 @@
+     */
+    @Override
+    public void execute(TaskList list, Ui ui, Storage storage) throws DukeException {
+        boolean markedDone = list.markDoneAtIndex(index);
```
Maybe isMarekedDone is a better boolean variable name in this case :)

ZhaoPeiduo commented on src/main/java/duke/AddCommand.java:
```diff
@@ -0,0 +1,47 @@
+     * @return True if it is a terminating command and false otherwise.
+     */
+    @Override
+    public boolean isExit() {
```
Maybe getIsExit() is a better choice for this method name, as isExit sounds like a boolean name to me.

ZhaoPeiduo replied:
EDIT: My bad your original method name follows the correct naming convention for boolean methods.

ZhaoPeiduo commented on src/main/java/duke/Command.java:
```diff
@@ -0,0 +1,26 @@
+     *
+     * @return True if it is a terminating command and false otherwise.
+     */
+    public abstract boolean isExit();
```
Similar issue as in AddCommand.class, maybe you need to make the change for Command and its subclasses.

ZhaoPeiduo replied:
EDIT: My bad your original method name follows the correct naming convention for boolean methods.

ZhaoPeiduo commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,57 @@
+ * @author Sherman Ng Wei Sheng
+ */
+public class Duke {
+    private final Storage storage;
+    private final Ui ui;
```
I like the use of final keyword here to indicate that storage and ui wouldnt be further modified.

shermannws committed:
Implement A-Gradle

ZhaoPeiduo commented on src/main/java/duke/FindCommand.java:
```diff
@@ -0,0 +1,62 @@
+     */
+    @Override
+    public boolean isExit() {
+        return false;
```
Maybe replace false with isExit (even though they are always the same)

shermannws committed:
Revert "Implement A-Gradle"

This reverts commit 8149f7610d58b24c82379003a1f8bd42d0514b09.

shermannws committed:
Test Commit

shermannws committed:
Implement A-Gradle again

shermannws committed:
Merge branch 'branch-A-Gradle'

ZhaoPeiduo reviewed (commented):
There are only few violations for java coding standard, as I have pointed out in the comments.
Well done!
(PS: The structure of different components of Duke is really clear and demonstrates strong OOP skills, I have personally learnt some tips from your code :))

ZhaoPeiduo commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,72 @@
+        try {
+            File storageDir = new File(this.fileDir);
+            File storageFile = new File(this.filePath);
+            boolean isDirExistent = storageDir.exists();
```
isDirExistent sounds a bit akward for me, maybe isDirPresent is a better alternative?

ZhaoPeiduo commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,77 @@
+    public void printAddedTaskMessage(Task task, TaskList list) {
+        String message =
+                "Got it. I've added this task:\n" +
+                "  " + task + "\n" +
```
According to the coding standard, the operator should be at the beginning of the new line. 
<img width="737" alt="屏幕截图 2021-08-31 220809" src="https://user-images.githubusercontent.com/77187494/131517643-2c0ea0e7-7050-4f96-ab98-772d9ffc7622.png">

shermannws committed:
Implement A-CheckStyle and fix styling errors.

shermannws committed:
Merge branch 'branch-A-CheckStyle'

yunpeng1234 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,56 @@
+        while (!isExit) {
+            try {
+                String input = this.ui.getInput();
+                Command c = Parser.parse(input);
```
Command C can be renamed to reflect it better. Something like ParsedCommand would be more helpful to the readers

yunpeng1234 commented on src/main/java/duke/MarkDoneCommand.java:
```diff
@@ -0,0 +1,54 @@
+     */
+    @Override
+    public void execute(TaskList list, Ui ui, Storage storage) throws DukeException {
+        boolean markedDone = list.markDoneAtIndex(index);
```
MarkedDown does not sound like a boolean. Maybe isMarked/isMarkedDone is more appropriate

yunpeng1234 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,72 @@
+        try {
+            File storageDir = new File(this.fileDir);
+            File storageFile = new File(this.filePath);
+            boolean isDirExistent = storageDir.exists();
```
Maybe doesDirExist would be a better name for the boolean

yunpeng1234 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,72 @@
+            File storageDir = new File(this.fileDir);
+            File storageFile = new File(this.filePath);
+            boolean isDirExistent = storageDir.exists();
+            boolean isFileExistent = storageFile.exists();
```
Likewise, doesFileExist would fit the naming convention of this class

yunpeng1234 commented on src/test/java/duke/DeadlineTest.java:
```diff
@@ -0,0 +1,43 @@
+        assertEquals("D | 0 | testingDeadlineTaskOne | 2021-08-26",
+                new Deadline("testingDeadlineTaskOne",
+                        LocalDate.of(2021, 8, 26),
+                false)
```
Indentation of false here should be consistent with the rest of the test cases. +4 more spaces to the right

yunpeng1234 reviewed (commented):
In general the coding standards are strong adhered to minus some minor changes.

For code quality wise, Duke exceptions can be abstracted out better so you don't have to input the full line to print per case of IllegalFormatting etc.

neojxuan commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,161 @@
+        for (String line : lines) {
+            String[] split = line.split(" \\| ");
+            String taskType = split[0];
+            boolean isDone = !split[1].equals("0");
```
Would it be better to catch an array out of bounds exception here? Just in case the file in storage is not formatted properly and split contains less than three elements .

neojxuan commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,76 @@
+    }
+
+    /**
+     * Prints the log for task object being added
```
Missing full stop

neojxuan commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,161 @@
+     */
+    public boolean markDoneAtIndex(int index) {
+        if (index > this.size() - 1 || index < 0) {
+            return false;
```
Is it possible to just throw the InvalidIndexException here instead of returning false?

neojxuan commented on src/main/java/duke/FindCommand.java:
```diff
@@ -0,0 +1,60 @@
+        TaskList newList = new TaskList();
+        for (int i = 0; i < size; i++) {
+            Task current = list.get(i);
+            String currentDescription = current.getDescription();
```
Could this be refactored so that a getter is not needed? Perhaps the Task class can have a containsKeyword function that checks if the given keyword is in the description.

neojxuan commented on src/main/java/duke/MarkDoneCommand.java:
```diff
@@ -0,0 +1,54 @@
+    @Override
+    public void execute(TaskList list, Ui ui, Storage storage) throws DukeException {
+        boolean markedDone = list.markDoneAtIndex(index);
+        if (markedDone) {
```
Similar to the other comment about throwing the exception directly, I believe this boolean value may not be necessary if the exception can be thrown by markDoneAtIndex.

neojxuan commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,132 @@
+        String[] splitWord = input.split(" ", 2);
+        String firstWord = splitWord[0];
+
+        if (input.equals("bye")) {
```
Do correct me if I'm wrong, but you should be able to use firstWord here as well rather than input. Then you can convert the if-else blocks to case statements which are slightly shorter/may be more readable. However, this is up to your preference haha

neojxuan commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,132 @@
+            String toDoDescription = input.split(" ", 2)[1];
+            if (toDoDescription.length() == 0) {
+                throw new IllegalFormatException("☹ OOPS!!! The description of a todo cannot be empty.");
+            }
```
Would it be possible to abstract these two checks into helper functions? For example, the code here could then be condensed into something like ```checkNonEmptyDescription(input, "todo");```. This can be reused in the blocks for other tasks like events and deadlines as well.

neojxuan commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,132 @@
+                deadlineDateTime = LocalDate.parse(splitTask[1]);
+            } catch (DateTimeParseException e) {
+                throw new IllegalFormatException("☹ OOPS!!! Specify the deadline date in the format yyyy-mm-dd.");
+            }
```
Could this be into another method like parseDateTime that can be reused for both deadlines and events?

neojxuan commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,132 @@
+            int secondWordLength = secondWord.length();
+            for (int i = 0; i < secondWordLength; i++) {
+                if (!Character.isDigit(secondWord.charAt(i))) {
+                    throw new IllegalFormatException("☹ OOPS!!! Specify the task to be mark done as a number.");
```
Error message should read task to  be deleted rather than to be marked done

neojxuan commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,132 @@
+            String secondWord = splitCommand[1];
+            int secondWordLength = secondWord.length();
+            for (int i = 0; i < secondWordLength; i++) {
+                if (!Character.isDigit(secondWord.charAt(i))) {
```
Since something similar is being done in both the done and delete commands, would it be better to abstract some parts into methods?

neojxuan commented on src/main/java/duke/Command.java:
```diff
@@ -0,0 +1,26 @@
+     *
+     * @return True if it is a terminating command and false otherwise.
+     */
+    public abstract boolean isExit();
```
Perhaps the default value returned by isExit() can be false? Assuming there will be fewer terminating commands than non-terminating ones. Then in the terminating commands, this function will be overridden, otherwise there is no need to override.

neojxuan commented on src/main/java/duke/AddCommand.java:
```diff
@@ -0,0 +1,46 @@
+ * @author Sherman Ng Wei Sheng
+ */
+public class AddCommand extends Command {
+    private boolean isExit;
```
What is the use of the isExit attribute here? Would it be possible to remove the attribute and just return false in isExit?

neojxuan reviewed (commented):
I think your code is generally very clean, clear and easy to understand. 😀 It also follows the Coding Standard very closely. I gave some minor suggestions for refactoring.

shermannws committed:
Implement Level-10

shermannws committed:
Merge branch 'branch-Level-10'

shermannws committed:
Fix Checkstyle Error for v2.0.0 release.

Code encounters with Checktyle error when building with Gradle.
Hence, code editing is being done to ensure that v2.0.0 release can be deployed.

shermannws committed:
Implement A-Assertions

Assertions are inserted to ensure that the program work desirably.
The following situations are being handled.
* Execution of AddCommand expect size of list to increase by 1.
* Execution of DeleteCommand expect size of list to decrease by 1.
* Execution of MarkDoneCommand expect size of list to remain the same.
* Marking a taskDone should expect the private isDone field to be true.
* Loading of taskList from Storage should only expect 3 types (T,D,E).

Some other minor fixes:
* Javadoc for Launcher.java.

shermannws committed:
Edit .gitignore

.gitignore should not be tracking the data directory which should be
unique to every local computer.

shermannws committed:
Edit .gitignore in branch

shermannws committed:
Implement A-CodeQuality

Examine & Refactor code and ensure that code quality is achieved.
SLAP is practise in almost all ways that are possible.

shermannws committed:
Merge pull request #1 from shermannws/branch-A-Assertions

Implement A-Assertions

shermannws committed:
Merge branch 'master' into branch-A-CodeQuality

shermannws committed:
Merge pull request #2 from shermannws/branch-A-CodeQuality

Implement A-CodeQuality

shermannws committed:
Implement C-Help

User is unable to know the full range of features that Duke is able to
offer.

Adding a help feature can allow users to check out DUke and its
functionalities.

The entire documentation is stored in a txt file in predetermined file
directory that storage will load and render it on the gui.

It is done this way so that future revision of the features will not
require programmers to edit the code, but instead, just have to edit the
txt file stored in [root]/data/documentation.txt

CHanges to other parts of the code was also being done to accommodate
for the revision in the DukeException class.

.git ignore has also been updated to also include the
data/documentation.txt file in the version control but not the duke.txt
file.

shermannws committed:
Merge branch 'branch-C-Help'

shermannws committed:
Enhance CodeQuality as per A-CodeQuality

Object fields have been accessed using "this" and does not align with
the Java convention.

This cause certain codes to be unnecessary long.

The this keywords at relevant parts of the code has been removed to
abide by the convention. However, take note that the this keyword has
not been removed from the constructors.

This ensures alignment with convention and readability of code.

shermannws committed:
Implement A-BetterGui

To appeal to the users of this program, the GUI has been designed to be
more attractive and at the same time, fixed certain bugs that did not
allow the program to work as desired.

Things that has been done
*GUI initialises the window with a welcome message
*GUI will not truncate the text if it exceeds the preset height
*Amended the format to be printed on GUI and on console

It is done this way to ensure that the program behaves ideally.

shermannws committed:
Fix gui send button alignment to bottom

shermannws committed:
Refactor and organize code

shermannws committed:
Add gui window title

shermannws committed:
Add Ui.png

shermannws committed:
Change Ui.png in docs folder

shermannws committed:
A-UserGuide

shermannws committed:
Revert "A-UserGuide"

This reverts commit 552a1c0dc3e04e896a0ae70b798434d00245f82d.

shermannws committed:
Revert "Revert "A-UserGuide""

This reverts commit d5dff27556203d4d99014f856571ff08f19fc70c.

shermannws committed:
Implement A-UserGuide

shermannws committed:
Set theme jekyll-theme-cayman

shermannws committed:
Edit header for README.md

shermannws committed:
Refactor code

Duke stores the string to be rendered when HelpCommand is executed in a
seprate file. However, when deployed to jar file, this txt file cannot
be read.

Let's create a separate class called HelpProvider so that the String
can still be retrieved when the jar file is used in another directory.

shermannws committed:
Refactor code

shermannws committed:
Refactor and repackage code

shermannws committed:
Refactor code for A-Release

A clean code is important for other developers to view this code.

Updating the javadoc to ensures that other developers are able to
understand the code.

Additional prompt should be available for users when they face with an
error message.

Let's,
* refactor the program to ensure that the code is ready for other
developers to view and to grade
* ensure that users are being prompted to the help page when an
error occurs.
